src/App.tsx

```diff
@@ -15,47 +16,116 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+type ReorderOptions = {
+  sortType: SortType,
+  isReversed: boolean,
+};
+
+const getReorderedGoods = (
+  goods: string[], { sortType, isReversed }: ReorderOptions,
+) => {
+  // To avoid the original array mutation
+  const visibleGoods = [...goods];
+
+  // Sort and reverse goods if needed
+  // eslint-disable-next-line no-console
+  console.log(sortType, isReversed);
+
+  visibleGoods.sort((good1, good2) => {
+    if (sortType === SortType.ALPHABET) {
+      return good1.localeCompare(good2);
+    }
+
+    if (sortType === SortType.LENGTH) {
+      return good1.length - good2.length;
+    }
+
+    return 0;
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+};
+
 export const App: React.FC = () => {
+  const [isReversed, setIsReversed] = useState(false);
+  const [sortType, setSortType] = useState<SortType>(SortType.NONE);
+  const goods = getReorderedGoods(
+    [...goodsFromServer], { sortType, isReversed },
+  );
+
+  const sort = (text: SortType) => {
+    setSortType(text);
+  };
```

Comment:
Here you can rewrite as Paul showed with carrying:
```
const handleSort = (type: SortType) => () => {
    setSortType(type);
  };
```
and use it in this way onClick={handleSort(SortType.ALPHABET)}, without callback